File: python/2019/01/aoc_2019_01.py
import sys
from common import timer, Puzzle


class Day1(Puzzle, year=2019, day=1):
    @staticmethod
    def clean_input(data):
        return [int(d) for d in data]

    @staticmethod
    def calc_fuel(mass):
        fuel = int(mass / 3) - 2
        if fuel < 0:
            return 0
        else:
            return fuel + Day1.calc_fuel(fuel)

    @timer(part=1)
    def solve_part_1(self):
        """Solution for part 1"""
        return sum([(int(m/3) - 2) for m in self.data])

    @timer(part=2)
    def solve_part_2(self):
        """Solution for part 2"""
        sum_fuel = 0
        for mass in self.data:
            fuel = int(mass / 3) - 2
            sum_fuel += fuel + Day1.calc_fuel(fuel)

        return sum_fuel


if __name__ == "__main__":
    filename = "input.txt"
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    Day1(filename=filename).solve_all()

File: C++/2019/01/aoc_2019_01.cpp
#include "aoc_io.hpp"
#include "aoc_timer.hpp"
#include <numeric>
#include <ranges>
#include <string>
#include <vector>

static int fuel_cost(int fuel) {
    fuel = (fuel / 3) - 2;
    if (fuel < 0) {
        return 0;
    }
    return fuel + fuel_cost(fuel);
}

std::vector<int> clean_input(std::vector<std::string> instructions) {
    std::vector<int> return_instructions;
    for (auto &row : instructions) {
        return_instructions.push_back(std::stoi(row));
    }

    return return_instructions;
}

int solve_part_1(std::vector<int> instructions) {

    auto fuel = instructions | std::views::transform([](const int &mass) -> int { return mass / 3 - 2; });

    auto result = std::reduce(fuel.begin(), fuel.end());
    return result;
}

int solve_part_2(std::vector<int> instructions) {

    auto fuel = instructions | std::views::transform([](const int &mass) -> int { return fuel_cost(mass); });

    auto result = std::reduce(fuel.begin(), fuel.end());
    return result;
}

std::string solve_all(std::vector<std::string> instructions) {
    std::vector<int> parsed = clean_input(instructions);
    aoc::timer<int>(1, solve_part_1, parsed, true);
    aoc::timer<int>(2, solve_part_2, parsed, true);

    return std::string("");
}

int main(int argc, char **argv) {
    std::string filename;
    const int   year = 2019;
    const int   day = 1;

    std::vector<std::string> instructions;

    if (argc > 1) {
        if (std::string(argv[1]) == "--test") {
            filename = "test_input.txt";
        } else {
            filename = argv[1];
        }
    } else {
        filename = "input.txt";
    }

    instructions = aoc::io::get_input_list<std::string>(filename, year, day);

    std::cout << "Solution for " << std::format("{:d}/{:02d}", year, day) << std::endl;
    aoc::timer<std::string>(0, solve_all, instructions, false);

    return 0;
}
